internal/clock/simulated_clock_test.go

```diff
@@ -0,0 +1,279 @@
+func assertNotReceivesTime(t *testing.T, ch <-chan time.Time, timeout time.Duration, msgAndArgs ...interface{}) {
+	t.Helper()
+	select {
+	case receivedTime := <-ch:
+		t.Fatalf("Expected no time on channel, but received %v. %s", receivedTime, msgAndArgs)
+	case <-time.After(timeout): // Using real time.After for test timeout
+		// Success, nothing received
+	}
+}
+
+func TestSimulatedClock_Now(t *testing.T) {
+	testCases := []struct {
+		name             string
+		initialTimeSetup func(sc *SimulatedClock)
+		expectedTime     time.Time
+	}{
+		{
+			name:             "InitialState_IsZeroTime",
+			initialTimeSetup: func(sc *SimulatedClock) {},
+			expectedTime:     referenceTime,
+		},
+		{
+			name: "AfterSetTime_ReturnsSetTime",
+			initialTimeSetup: func(sc *SimulatedClock) {
+				sc.SetTime(referenceTime)
+			},
+			expectedTime: referenceTime,
+		},
+		{
+			name: "AfterAdvanceTime_ReturnsAdvancedTime",
+			initialTimeSetup: func(sc *SimulatedClock) {
+				sc.SetTime(referenceTime)
+				sc.AdvanceTime(time.Hour)
+			},
+			expectedTime: referenceTime.Add(time.Hour),
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			clock := NewSimulatedClock(referenceTime)
+			tc.initialTimeSetup(clock)
+
+			now := clock.Now()
+
+			assert.True(t, now.Equal(tc.expectedTime), "clock.Now() returned %v, expected %v", now, tc.expectedTime)
+		})
+	}
+}
+
+func TestSimulatedClock_SetTime(t *testing.T) {
+	testCases := []struct {
+		name             string
+		initialTimeSetup func(sc *SimulatedClock) // To set a pre-existing time if needed
+		timeToSet        time.Time
+		expectedAfter    time.Time
+	}{
+		{
+			name:             "SetFromZeroTime",
+			initialTimeSetup: func(sc *SimulatedClock) {},
+			timeToSet:        referenceTime,
+			expectedAfter:    referenceTime,
+		},
+		{
+			name: "OverwriteExistingTime",
+			initialTimeSetup: func(sc *SimulatedClock) {
+				sc.SetTime(referenceTime.Add(-time.Hour)) // Start with a different time
+			},
+			timeToSet:     referenceTime,
+			expectedAfter: referenceTime,
+		},
+		{
+			name:             "SetToZeroTime",
+			initialTimeSetup: func(sc *SimulatedClock) { sc.SetTime(referenceTime) },
+			timeToSet:        time.Time{}, // Zero time
+			expectedAfter:    time.Time{},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			clock := NewSimulatedClock(referenceTime)
+			tc.initialTimeSetup(clock)
+
+			clock.SetTime(tc.timeToSet)
+
+			assert.True(t, clock.Now().Equal(tc.expectedAfter), "After SetTime, clock.Now() was %v, expected %v", clock.Now(), tc.expectedAfter)
+		})
+	}
+}
+
+func TestSimulatedClock_AdvanceTime(t *testing.T) {
+	testCases := []struct {
+		name         string
+		initialTime  time.Time
+		advanceBy    time.Duration
+		expectedTime time.Time
+	}{
+		{
+			name:         "AdvancePositiveDuration",
+			initialTime:  referenceTime,
+			advanceBy:    5 * time.Minute,
+			expectedTime: referenceTime.Add(5 * time.Minute),
```

Comment:
Can we make tests damp and dumb and not have logic in them - https://testing.googleblog.com/2014/07/testing-on-toilet-dont-put-logic-in.html and https://testing.googleblog.com/2019/12/testing-on-toilet-tests-too-dry-make.html.

I'd encourage to get rid of referenceTime and use actual time throughout.